### PR TITLE
Add integration tests for CDC event conversion from logical replication stream with composite types

### PR DESCRIPTION
**Pull request by peterxcli**
## Summary

Add integration tests for CDC event conversion from logical replication stream with composite types

## Related Issues

Closes https://github.com/Mooncake-Labs/moonlink/issues/1524

## Changes

- Add pg cdc event IT skeleton and one `test_composite_types_in_cdc_stream` test case to test   **composite**, **array of composite**, **nested composite** and **nested array in composite**
- Add a `TestResources` struct with drop implemented to release pg resource
- Correct pg wal level config correctly in github ci file

## Checklist

- [x] Code builds correctly
- [x] Tests have been added or updated
- [x] Documentation updated if necessary
- [x] I have reviewed my own changes

## Notes

```
Event 14: XLogData(XLogDataBody { wal_start: 54529448, wal_end: 54529448, timestamp: 808677494744232, data: Update(UpdateBody { xid: None, rel_id: 28017, old_tuple: None, key_tuple: None, new_tuple: Tuple([Text(b"1"), Text(b"(\"Updated St\",\"Updated City\",54321)"), Text(b"(Home,\"(40.7,-74)\")"), Text(b"(Alice,\"{\"\"(\\\\\"\"123 Main St\\\\\"\",NYC,10001)\"\",\"\"(\\\\\"\"456 Oak Ave\\\\\"\",LA,90210)\"\"}\",\"(Work,\"\"(34,-118)\"\")\")"), Text(b"{\"(\\\"789 Pine St\\\",Chicago,60601)\",\"(\\\"321 Elm St\\\",Boston,2101)\"}")]) }) })
```

### CODE REVIEW

**Comment by peterxcli** [comment]
??
this should fail with:
```bash
running 1 test
Error in CDC stream: CdcEventConversion(FromBytes(InvalidComposite(MissingParentheses)))
Error in CDC stream: CdcEventConversion(FromBytes(InvalidComposite(MissingParentheses)))

thread 'pg_replicate::tests::cdc_event_test::test_composite_types_in_cdc_stream' panicked at src/moonlink_connectors/src/pg_replicate/tests/cdc_event_test.rs:374:10:
expected insert for id=3
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
test pg_replicate::tests::cdc_event_test::test_composite_types_in_cdc_stream ... FAILED

failures:

failures:
    pg_replicate::tests::cdc_event_test::test_composite_types_in_cdc_stream

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 79 filtered out; finished in 5.36s
```

without the patch(handle doubled quotes inside quoted composite/array fields):
```diff
diff --git a/src/moonlink_connectors/src/pg_replicate/conversions/text.rs b/src/moonlink_connectors/src/pg_replicate/conversions/text.rs
index 18be04d..464bb5c 100644
--- a/src/moonlink_connectors/src/pg_replicate/conversions/text.rs
+++ b/src/moonlink_connectors/src/pg_replicate/conversions/text.rs
@@ -383,7 +383,7 @@ impl TextFormatConverter {
         let mut in_quotes = false;
         let mut in_escape = false;
         let mut val_quoted = false;
-        let mut chars = str.chars();
+        let mut chars = str.chars().peekable();
         let mut done = str.is_empty();
 
         while !done {
@@ -395,10 +395,19 @@ impl TextFormatConverter {
                             in_escape = false;
                         }
                         '"' => {
-                            if !in_quotes {
+                            if in_quotes {
+                                // support doubled quotes inside quoted value
+                                if let Some('"') = chars.peek().copied() {
+                                    // consume next quote and append a single quote to value
+                                    let _ = chars.next();
+                                    val_str.push('"');
+                                } else {
+                                    in_quotes = false;
+                                }
+                            } else {
                                 val_quoted = true;
+                                in_quotes = true;
                             }
-                            in_quotes = !in_quotes;
                         }
                         '\\' => in_escape = true,
                         ',' if !in_quotes => {
@@ -454,7 +463,7 @@ impl TextFormatConverter {
         let mut in_quotes = false;
         let mut in_escape = false;
         let mut val_quoted = false;
-        let mut chars = inner.chars();
+        let mut chars = inner.chars().peekable();
         let mut field_iter = fields.iter();
         let mut done = inner.is_empty();
 
@@ -467,10 +476,18 @@ impl TextFormatConverter {
                             in_escape = false;
                         }
                         '"' => {
-                            if !in_quotes {
+                            if in_quotes {
+                                // doubled quotes inside quoted composite field
+                                if let Some('"') = chars.peek().copied() {
+                                    let _ = chars.next();
+                                    val_str.push('"');
+                                } else {
+                                    in_quotes = false;
+                                }
+                            } else {
                                 val_quoted = true;
+                                in_quotes = true;
                             }
-                            in_quotes = !in_quotes;
                         }
                         '\\' if in_quotes => in_escape = true,
                         ',' if !in_quotes => {
@@ -547,7 +564,7 @@ impl TextFormatConverter {
         let mut val_str = String::with_capacity(50);
         let mut in_quotes = false;
         let mut in_escape = false;
-        let mut chars = inner.chars();
+        let mut chars = inner.chars().peekable();
         let mut done = false;
 
         while !done {
@@ -558,11 +575,18 @@ impl TextFormatConverter {
                             val_str.push(c);
                             in_escape = false;
                         }
-                        '\\' if in_quotes => {
-                            in_escape = true;
-                        }
+                        '\\' if in_quotes => in_escape = true,
                         '"' => {
-                            in_quotes = !in_quotes;
+                            if in_quotes {
+                                if let Some('"') = chars.peek().copied() {
+                                    let _ = chars.next();
+                                    val_str.push('"');
+                                } else {
+                                    in_quotes = false;
+                                }
+                            } else {
+                                in_quotes = true;
+                            }
                         }
                         ',' if !in_quotes => {
                             break;
```

at least on my local machine...

**Comment by peterxcli** [comment]
@dentiny could you help me to checkout to this patch and run `cargo test --package moonlink_connectors --lib --features connector-pg`? thanks!

File: src/moonlink_connectors/src/pg_replicate/tests/cdc_event_test.rs [hunk 1]
```diff
@@ -0,0 +1,460 @@
...
+        .await
+        .unwrap();
+    let slot_info = match tokio::time::timeout(
+        Duration::from_secs(10),
+        replication_client.get_or_create_slot(REPLICATION_SLOT),
+    )
+    .await
+    {
+        Ok(Ok(slot_info)) => slot_info,
+        Ok(Err(_)) => return,
```
Comment (dentiny) [inline src/moonlink_connectors/src/pg_replicate/tests/cdc_event_test.rs]: Do we return the whole function early if error?
I would suggest log or unwrap in tests, instead of silently proceed

**Comment by peterxcli** [comment]
> ?? this should fail with:
> 
> ```shell
> running 1 test
> Error in CDC stream: CdcEventConversion(FromBytes(InvalidComposite(MissingParentheses)))
> Error in CDC stream: CdcEventConversion(FromBytes(InvalidComposite(MissingParentheses)))

Sorry this was a pg setting in ci issue and the wrongly early return in test.
Not the error in ci is expected: https://github.com/Mooncake-Labs/moonlink/actions/runs/17018798508/job/48244935785?pr=1525#step:10:372
👍

**Comment by peterxcli** [comment]
Let's get https://github.com/Mooncake-Labs/moonlink/pull/1541 merged frist.
cc @dentiny

**Comment by dentiny** [comment]
> Let's get #1541 merged frist. cc @dentiny

Do you mind rebasing? Thanks!

**Comment by dentiny** [comment]
Ping me when ready, the chaos tests are still pretty broken, feel free to ignore it.

**Comment by peterxcli** [comment]
@dentiny This PR is ready for review, please take another look. Thanks!
But if you feel like this patch can be spitted to 1. TestResources + pg_type_test adapt TestResources. 2. pure cdc event integration test. I'm ok for that.

File: src/moonlink_connectors/src/pg_replicate/tests/cdc_event_test.rs [hunk 2]
```diff
@@ -0,0 +1,419 @@
...
+    let mut schema_client = ReplicationClient::from_client(schema_pg_client);
+    let table_name = TableName {
+        schema: "public".to_string(),
+        name: table_name_str.to_string(),
+    };
+    let src_table_id = schema_client
+        .get_src_table_id(&table_name)
+        .await
+        .unwrap()
+        .expect("missing table id");
```
Comment (dentiny) [inline src/moonlink_connectors/src/pg_replicate/tests/cdc_event_test.rs]:
```suggestion
        .expect("missing table id for table {table_name}");
```

File: src/moonlink_connectors/src/pg_replicate/tests/cdc_event_test.rs [hunk 3]
```diff
@@ -0,0 +1,419 @@
...
+    schema_client
+        .get_table_schema(src_table_id, table_name, Some(publication))
+        .await
+        .unwrap()
```
Comment (dentiny) [inline src/moonlink_connectors/src/pg_replicate/tests/cdc_event_test.rs]: curious what's the purpose for getting table schema here? We're not using it

File: src/moonlink_connectors/src/pg_replicate/tests/cdc_event_test.rs [hunk 4]
```diff
@@ -0,0 +1,419 @@
...
+}
+
+fn spawn_sql_executor(database_url: String) -> mpsc::UnboundedSender<String> {
```
Comment (dentiny) [inline src/moonlink_connectors/src/pg_replicate/tests/cdc_event_test.rs]: Let's add a doc

File: src/moonlink_connectors/src/pg_replicate/tests/test_utils.rs [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+            eprintln!("Replication connection error: {e}");
+        }
+    });
+    replication_client
+}
+
+async fn drop_publication(client: &tokio_postgres::Client, publication: &str) {
+    let _ = client
+        .simple_query(&format!(
+            "DROP PUBLICATION IF EXISTS {publication} CASCADE;"
```
Comment (dentiny) [inline src/moonlink_connectors/src/pg_replicate/tests/test_utils.rs]:
```suggestion
            "DROP PUBLICATION {publication} CASCADE;"
```
I'm thinking in the test, where we have fully controlled env, should we fail the test if expected publication and table doesn't exist? I'm afraid it could hide problems.

File: src/moonlink_connectors/src/pg_replicate/tests/test_utils.rs [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+        ))
+        .await
+        .unwrap();
+}
+
+async fn drop_table(client: &tokio_postgres::Client, table_name: &str) {
+    let _ = client
+        .simple_query(&format!("DROP TABLE IF EXISTS {table_name} CASCADE;"))
```
Comment (dentiny) [inline src/moonlink_connectors/src/pg_replicate/tests/test_utils.rs]:
```suggestion
        .simple_query(&format!("DROP TABLE {table_name} CASCADE;"))
```

File: src/moonlink_connectors/src/pg_replicate/tests/test_utils.rs [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+        // Drop the SQL sender to stop background executor if still running
+        let _ = self.sql_tx.take();
+
+        // Use Arc to share the client without cloning
+        let client = Arc::clone(&self.client);
+        let tables = self.tables.clone();
+        let publications = self.publications.clone();
+        let slots = self.slots.clone();
+        let types = self.types.clone();
+        tokio::spawn(async move {
```
Comment (dentiny) [inline src/moonlink_connectors/src/pg_replicate/tests/test_utils.rs]: I don't think this is ideal for unit test, which should be deterministic and self-contained.
One example you could follow is
https://github.com/Mooncake-Labs/moonlink/blob/86b1549be754345f179e03f4ae638a1e82939208/src/moonlink_backend/tests/common.rs#L104-L113
and you need to use multi-threads for all tests
https://github.com/Mooncake-Labs/moonlink/blob/86b1549be754345f179e03f4ae638a1e82939208/src/moonlink_backend/tests/test_backend.rs#L76

File: src/moonlink_connectors/src/pg_replicate/tests/cdc_event_test.rs [hunk 8]
```diff
@@ -0,0 +1,409 @@
...
+    });
+    tx
+}
+
+#[tokio::test]
+#[serial]
+async fn test_composite_types_in_cdc_stream() {
+    let client = setup_connection().await;
+    // Per-test dynamic names
+    let suffix = std::time::SystemTime::now()
```
Comment (dentiny) [inline src/moonlink_connectors/src/pg_replicate/tests/cdc_event_test.rs]: You have declare the unit test to be executed in serial, so no concurrent access to our pg instance; curious why do we need the random suffix?

File: src/moonlink_connectors/src/pg_replicate/tests/test_utils.rs [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+async fn drop_table(client: &tokio_postgres::Client, table_name: &str) {
+    let _ = client
+        .simple_query(&format!("DROP TABLE IF EXISTS {table_name} CASCADE;"))
+        .await
+        .unwrap();
+}
+
+async fn drop_type(client: &tokio_postgres::Client, type_name: &str) {
+    let _ = client
+        .simple_query(&format!("DROP TYPE IF EXISTS {type_name} CASCADE;"))
```
Comment (dentiny) [inline src/moonlink_connectors/src/pg_replicate/tests/test_utils.rs]:
```suggestion
        .simple_query(&format!("DROP TYPE {type_name} CASCADE;"))
```

File: src/moonlink_connectors/src/pg_replicate/tests/test_utils.rs [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+        .await
+        .unwrap();
+}
+
+async fn drop_replication_slot(client: &tokio_postgres::Client, slot_name: &str) {
+    let _ = client
+        .simple_query(&format!(
+            "SELECT pg_drop_replication_slot('{slot}');",
```
Comment (dentiny) [inline src/moonlink_connectors/src/pg_replicate/tests/test_utils.rs:143]: it reads `select` statement but you name it `drop_slot`?

File: src/moonlink_connectors/src/pg_replicate/tests/test_utils.rs [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+        self.types.push(name.into());
+    }
+
+    pub fn set_sql_tx(&mut self, tx: mpsc::UnboundedSender<String>) {
+        self.sql_tx = Some(tx);
+    }
+
+    pub async fn cleanup(&self) {
+        for slot in &self.slots {
+            drop_replication_slot(&self.client, slot).await;
```
Comment (dentiny) [inline src/moonlink_connectors/src/pg_replicate/tests/test_utils.rs]: I'm actually curious, what's the motivation of having both clean up function and drop?
At least we could reuse the same cleanup?

File: src/moonlink_connectors/src/pg_replicate/tests/test_utils.rs [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+
+    pub fn add_type(&mut self, name: impl Into<String>) {
+        self.types.push(name.into());
+    }
+
+    pub fn set_sql_tx(&mut self, tx: mpsc::UnboundedSender<String>) {
+        self.sql_tx = Some(tx);
+    }
+
+    pub async fn cleanup(&self) {
```
Comment (dentiny) [inline src/moonlink_connectors/src/pg_replicate/tests/test_utils.rs]:
```suggestion
    pub async fn cleanup(self) {
```

File: src/moonlink_connectors/src/pg_replicate/tests/pg_type_tests.rs [hunk 13]
```diff
@@ -59,13 +46,17 @@ async fn test_basic_composite_type() {
         .await
         .unwrap();
 
-    let table_id = get_table_id(&client, "test_basic_composite").await;
+    // Register resources for cleanup
+    resources.add_type("test_address");
+    resources.add_table("test_basic_composite");
```
Comment (dentiny) [inline src/moonlink_connectors/src/pg_replicate/tests/pg_type_tests.rs]: let's make both type and table constants, it's hard to tell what entity it is from their names.

File: src/moonlink_connectors/src/pg_replicate/tests/pg_type_tests.rs [hunk 14]
```diff
@@ -119,8 +112,13 @@ async fn test_nested_composite_types() {
         .await
         .unwrap();
 
-    let table_id = get_table_id(&client, "test_nested").await;
-    let replication_client = ReplicationClient::from_client(client);
+    // Register resources for cleanup
+    resources.add_type("test_point");
```
Comment (dentiny) [inline src/moonlink_connectors/src/pg_replicate/tests/pg_type_tests.rs]: ditto

File: src/moonlink_connectors/src/pg_replicate/tests/pg_type_tests.rs [hunk 15]
```diff
@@ -180,8 +180,12 @@ async fn test_array_of_composite_types() {
         .await
         .unwrap();
 
-    let table_id = get_table_id(&client, "test_array_composite").await;
-    let replication_client = ReplicationClient::from_client(client);
+    // Register resources for cleanup
```
Comment (dentiny) [inline src/moonlink_connectors/src/pg_replicate/tests/pg_type_tests.rs:230]: ditto

File: src/moonlink_connectors/src/pg_replicate/tests/pg_type_tests.rs [hunk 16]
```diff
@@ -249,8 +254,13 @@ async fn test_composite_with_nested_array() {
         .await
         .unwrap();
 
-    let table_id = get_table_id(&client, "test_nested_array").await;
-    let replication_client = ReplicationClient::from_client(client);
+    // Register resources for cleanup
+    resources.add_type("test_skill");
```
Comment (dentiny) [inline src/moonlink_connectors/src/pg_replicate/tests/pg_type_tests.rs]: ditto

File: src/moonlink_connectors/src/pg_replicate/tests/pg_type_tests.rs [hunk 17]
```diff
@@ -328,8 +339,15 @@ async fn test_deeply_nested_composites() {
         .await
         .unwrap();
 
-    let table_id = get_table_id(&client, "test_deep").await;
-    let replication_client = ReplicationClient::from_client(client);
+    // Register resources for cleanup
+    resources.add_type("test_coords");
```
Comment (dentiny) [inline src/moonlink_connectors/src/pg_replicate/tests/pg_type_tests.rs]: ditto

File: src/moonlink_connectors/src/pg_replicate/tests/pg_type_tests.rs [hunk 18]
```diff
@@ -415,8 +434,14 @@ async fn test_complex_mixed_nesting() {
         .await
         .unwrap();
 
-    let table_id = get_table_id(&client, "test_mixed").await;
-    let replication_client = ReplicationClient::from_client(client);
+    // Register resources for cleanup
+    resources.add_type("test_tag");
```
Comment (dentiny) [inline src/moonlink_connectors/src/pg_replicate/tests/pg_type_tests.rs]: ditto

File: src/moonlink_connectors/src/pg_replicate/tests/cdc_event_test.rs [hunk 19]
```diff
@@ -0,0 +1,409 @@
...
+                eprintln!("background SQL failed: {e}; sql: {sql}");
+            }
+        }
+    });
+    tx
+}
+
+#[tokio::test]
+#[serial]
+async fn test_composite_types_in_cdc_stream() {
```
Comment (dentiny) [inline src/moonlink_connectors/src/pg_replicate/tests/cdc_event_test.rs]: nitpick: do you mind adding another test case where there's stringified "null", so we could cover the bug fix at https://github.com/Mooncake-Labs/moonlink/pull/1541?
I assume it's much easier to setup in sql level than manually crafted pg types.

File: src/moonlink_connectors/src/pg_replicate/tests/cdc_event_test.rs [hunk 20]
```diff
@@ -0,0 +1,409 @@
...
+             );"
+        ))
+        .await
+        .unwrap();
+
+    let mut replication_client = create_replication_client().await;
+
+    // Create replication slot using the replication client
+    replication_client
+        .begin_readonly_transaction()
```
Comment (dentiny) [inline src/moonlink_connectors/src/pg_replicate/tests/cdc_event_test.rs]: (not a merge blocker) this is sth I'm not sure about, why do we open a read-only transaction, but create a replication slot inside (which looks like a write operation)?

File: src/moonlink_connectors/src/pg_replicate/tests/cdc_event_test.rs [hunk 21]
```diff
@@ -0,0 +1,409 @@
...
+    let table_schema = fetch_table_schema(&publication, &table_name).await;
+    use std::pin::Pin;
+    let mut pinned_stream = Box::pin(cdc_stream);
+    pinned_stream.as_mut().add_table_schema(table_schema);
+
+    // Start a background SQL executor and submit changes via channel
+    let sql_tx = spawn_sql_executor(database_url());
+    resources.set_sql_tx(sql_tx.clone());
+    // Developers can submit arbitrary SQL in desired order with optional delays between them
+    tokio::time::sleep(Duration::from_millis(100)).await;
```
Comment (dentiny) [inline src/moonlink_connectors/src/pg_replicate/tests/cdc_event_test.rs]: Actually why do we need the timeout between executions? we're blocking receiving their cdc events anyway.

File: src/moonlink_connectors/src/pg_replicate/tests/cdc_event_test.rs [hunk 22]
```diff
@@ -0,0 +1,409 @@
...
+            assert!(matches!(fields.get(0), Some(Cell::String(s)) if s == "999 Test St"));
+            assert!(matches!(fields.get(1), Some(Cell::String(s)) if s == "Test City"));
+            assert!(matches!(fields.get(2), Some(Cell::I32(12345))));
+        }
+        other => panic!("unexpected basic_addr cell: {:?}", other),
+    }
+
+    // Optionally validate update for id=1 if present
+    if let Some(updated_row) = events.iter().find_map(|e| {
+        if let CdcEvent::Update((_, _old, new_row, _)) = e {
```
Comment (dentiny) [inline src/moonlink_connectors/src/pg_replicate/tests/cdc_event_test.rs]:
```suggestion
        if let CdcEvent::Update((_, _, new_row, _)) = e {
```

File: src/moonlink_connectors/src/pg_replicate/tests/cdc_event_test.rs [hunk 23]
```diff
@@ -0,0 +1,409 @@
...
+            }
+        }
+    }
+
+    assert!(!events.is_empty(), "No CDC events were received");
+
+    use crate::pg_replicate::conversions::cdc_event::CdcEvent;
+
+    // the events might include other types of events, eg. BEGIN, PrimaryKeepAlive, etc.
+    // Filter & find insert for id=2
```
Comment (dentiny) [inline src/moonlink_connectors/src/pg_replicate/tests/cdc_event_test.rs]: just a nitpick (not a merge blocker), we're searching interesting events every time, so even if events are arriving the incorrect order tests will still pass; might just use a deque and pop front while search.

File: src/moonlink_connectors/src/pg_replicate/tests/cdc_event_test.rs (hunk 20 above)
Comment (peterxcli) [reply src/moonlink_connectors/src/pg_replicate/tests/cdc_event_test.rs]: Actually I have the exact same question in the prod code:
https://github.com/Mooncake-Labs/moonlink/blob/86b1549be754345f179e03f4ae638a1e82939208/src/moonlink_connectors/src/pg_replicate/clients/postgres.rs#L764-L766

https://github.com/Mooncake-Labs/moonlink/blob/86b1549be754345f179e03f4ae638a1e82939208/src/moonlink_connectors/src/pg_replicate/clients/postgres.rs#L730-L753

I think we should use `BEGIN isolation level repeatable read` at [line 765](https://github.com/Mooncake-Labs/moonlink/blob/86b1549be754345f179e03f4ae638a1e82939208/src/moonlink_connectors/src/pg_replicate/clients/postgres.rs#L765)

Comment (dentiny) [reply src/moonlink_connectors/src/pg_replicate/tests/cdc_event_test.rs]: Sorry I don't have too much context and knowledge here; asked in team internal channel (simply rephrase and copy it here).

Two questions you might be curious about:
- why read-only transaction for write operations?
- why no commit after begin transaction in the same function?

The first answer is: https://github.com/supabase/etl/blob/4da956c6b9be8476a1dbe87a4d88689e0671b7c1/etl/docs/Replication%20in%20Postgres.md?plain=1#L70
The second answers is: due to the way we implement replication and initial copy, we split commit begin and commit in two places, likely will be refactored later
https://github.com/Mooncake-Labs/moonlink/pull/918#discussion_r2218087961

**Comment by peterxcli** [comment]
Status update: addressed review comments and ready for re-review ✅

- Doubled-quote parsing: Implemented for text arrays, composites, and composite arrays in `src/moonlink_connectors/src/pg_replicate/conversions/text.rs` (uses `chars().peekable()` and consumes doubled quotes inside quoted fields).
- CI for Postgres: Workflow config now sets `wal_level=logical` and restarts the container before tests; this resolves the earlier mismatch noted in discussion.
- Integration tests: Added behind `connector-pg` feature in:
  - `src/moonlink_connectors/src/pg_replicate/tests/{test_utils.rs, pg_type_tests.rs, cdc_event_test.rs}`
  - Tests use `TestResources` for dynamic names + cleanup; CDC test covers composite, array-of-composite, nested composite, and arrays inside composites.
- Build/lint: `cargo fmt --check` and `cargo clippy --all-targets --all-features -- -D warnings` are clean.
- Compile check for reviewer ask: `cargo test --package moonlink_connectors --lib --features connector-pg --no-run` succeeds.
- Running the DB-backed tests: I can’t execute them in this environment; CI does via the new `PostgreSQL Integration Tests` job. Locally, run with `DATABASE_URL=postgresql://postgres:postgres@localhost:5432/postgres`.
- Rebase note: Rebased on `main` after #1541; includes the composite parsing fix.

If you prefer, I can split this into:
1) `TestResources` + adapting `pg_type_tests` and
2) CDC event integration test
—but kept together here since both are green and interrelated. Let me know your preference.

Ping @dentiny — ready for another look; ignoring chaos tests per your guidance.

File: src/moonlink_connectors/src/pg_replicate/tests/cdc_event_test.rs (hunk 20 above)
Comment (peterxcli) [reply src/moonlink_connectors/src/pg_replicate/tests/cdc_event_test.rs]: raised one - https://github.com/Mooncake-Labs/moonlink/issues/1588
for the second Q/A

File: src/moonlink_connectors/src/pg_replicate/tests/test_utils.rs (hunk 10 above)
Comment (peterxcli) [reply src/moonlink_connectors/src/pg_replicate/tests/test_utils.rs:143]: This is how `pg_drop_replication_slot` is used.
ref: https://stackoverflow.com/a/30857037

File: src/moonlink_connectors/src/pg_replicate/tests/cdc_event_test.rs (hunk 3 above)
Comment (peterxcli) [reply src/moonlink_connectors/src/pg_replicate/tests/cdc_event_test.rs]: We have to give CdcStream the schema of that table, otherwise it wont be able to parse the cdc content.
```rust
let table_schema = fetch_table_schema(&publication, &table_name).await;
    use std::pin::Pin;
    let mut pinned_stream = Box::pin(cdc_stream);
    pinned_stream.as_mut().add_table_schema(table_schema);
```

Comment (peterxcli) [reply src/moonlink_connectors/src/pg_replicate/tests/cdc_event_test.rs]: https://github.com/Mooncake-Labs/moonlink/blob/f0606173411498be24bbe87081463348b6fc5e6b/src/moonlink_connectors/src/pg_replicate/postgres_source.rs#L338-L344

https://github.com/Mooncake-Labs/moonlink/blob/f0606173411498be24bbe87081463348b6fc5e6b/src/moonlink_connectors/src/pg_replicate/postgres_source.rs#L363-L366

File: src/moonlink_connectors/src/pg_replicate/tests/cdc_event_test.rs (hunk 8 above)
Comment (peterxcli) [reply src/moonlink_connectors/src/pg_replicate/tests/cdc_event_test.rs]: yeah, ur right

**Comment by peterxcli** [comment]
@dentiny I've address all of your comments, thanks again for you thorough review, please take another look, thanks!
and sorry for this big patch 😭

File: src/moonlink_connectors/src/pg_replicate/tests/cdc_event_test.rs (hunk 3 above)
Comment (dentiny) [reply src/moonlink_connectors/src/pg_replicate/tests/cdc_event_test.rs]: My question is less of pg, more of:
- this function is named as get_xxx, which serves for read-only
- if we don't need the return value, why do we call it?

Comment (peterxcli) [reply src/moonlink_connectors/src/pg_replicate/tests/cdc_event_test.rs]: isn't the `schema_client.get_table_schema(src_table_id, table_name, Some(publication)).await.unwrap()` return the value of `fetch_table_schema`?
```rust
async fn fetch_table_schema(publication: &str, table_name_str: &str) -> TableSchema {
    // ...

    // this will return the schema of the given table 
    schema_client
        .get_table_schema(src_table_id, table_name, Some(publication))
        .await
        .unwrap()
}
```

Comment (dentiny) [reply src/moonlink_connectors/src/pg_replicate/tests/cdc_event_test.rs]: Ahhh, sorry I must be mis-reading... sorry!

File: src/moonlink_connectors/src/pg_replicate/tests/cdc_event_test.rs [hunk 24]
```diff
@@ -0,0 +1,511 @@
...
+    let url = database_url();
+    let (schema_pg_client, schema_conn) = connect(&url, NoTls).await.unwrap();
+    tokio::spawn(async move {
+        if let Err(e) = schema_conn.await {
+            eprintln!("Schema connection error: {e}");
+        }
+    });
+    let mut schema_client = ReplicationClient::from_client(schema_pg_client);
+    let table_name = TableName {
+        schema: "public".to_string(),
```
Comment (dentiny) [inline src/moonlink_connectors/src/pg_replicate/tests/cdc_event_test.rs]: nit: make it a constant

File: src/moonlink_connectors/src/pg_replicate/tests/cdc_event_test.rs [hunk 25]
```diff
@@ -0,0 +1,511 @@
...
+}
+
+/// Spawns a background SQL executor that can be used to submit arbitrary SQL in desired order with optional delays between them
+fn spawn_sql_executor(database_url: String) -> mpsc::UnboundedSender<String> {
+    let (tx, mut rx) = mpsc::unbounded_channel::<String>();
+    tokio::spawn(async move {
+        let (bg_client, bg_connection) = connect(&database_url, NoTls).await.unwrap();
+        tokio::spawn(async move {
+            if let Err(e) = bg_connection.await {
+                eprintln!("Background connection error: {e}");
```
Comment (dentiny) [inline src/moonlink_connectors/src/pg_replicate/tests/cdc_event_test.rs]: I'm thinking if we should directly fail it in unit test? Since it could actually block and timeout the whole test suite

File: src/moonlink_connectors/src/pg_replicate/tests/cdc_event_test.rs [hunk 26]
```diff
@@ -0,0 +1,511 @@
...
+            }
+        });
+
+        while let Some(sql) = rx.recv().await {
+            if let Err(e) = bg_client.simple_query(&sql).await {
+                eprintln!("background SQL failed: {e}; sql: {sql}");
```
Comment (dentiny) [inline src/moonlink_connectors/src/pg_replicate/tests/cdc_event_test.rs]: same here

File: src/moonlink_connectors/src/pg_replicate/tests/cdc_event_test.rs [hunk 27]
```diff
@@ -0,0 +1,511 @@
...
+              ARRAY[ROW('789 Pine St', 'Chicago', 60601)::test_address,
+                    ROW('321 Elm St', 'Boston', 02101)::test_address]
+             );"
+        ))
+        .await
+        .unwrap();
+
+    let mut replication_client = create_replication_client().await;
+
+    // Create replication slot using the replication client
```
Comment (dentiny) [inline src/moonlink_connectors/src/pg_replicate/tests/cdc_event_test.rs]: Do you mind if we putting the link besides?
https://github.com/supabase/etl/blob/4da956c6b9be8476a1dbe87a4d88689e0671b7c1/etl/docs/Replication%20in%20Postgres.md?plain=1#L70
just for better clarity

File: src/moonlink_connectors/src/pg_replicate/tests/cdc_event_test.rs [hunk 28]
```diff
@@ -0,0 +1,511 @@
...
+    let table_name = format!("test_cdc_string_null");
+    let publication = format!("test_cdc_string_null_pub");
+    let slot_name = format!("test_cdc_string_null_slot");
+
+    let mut resources = TestResources::new(client);
+    resources.add_table(table_name.clone());
+    resources.add_publication(publication.clone());
+    resources.add_slot(slot_name.clone());
+
+    // Simple schema with text columns to verify stringified "null"
```
Comment (dentiny) [inline src/moonlink_connectors/src/pg_replicate/tests/cdc_event_test.rs:339]: quite a few test setups could be extracted into common test utils, including replica identity, creation publication, create slot, etc.

File: src/moonlink_connectors/src/pg_replicate/tests/test_utils.rs (hunk 10 above)
Comment (dentiny) [reply src/moonlink_connectors/src/pg_replicate/tests/test_utils.rs:143]: TIL, thank you!

File: src/moonlink_connectors/src/pg_replicate/tests/cdc_event_test.rs [hunk 29]
```diff
@@ -46,15 +47,11 @@ fn spawn_sql_executor(database_url: String) -> mpsc::UnboundedSender<String> {
...
-                eprintln!("Background connection error: {e}");
-            }
+            bg_connection.await.unwrap();
         });
 
         while let Some(sql) = rx.recv().await {
-            if let Err(e) = bg_client.simple_query(&sql).await {
-                eprintln!("background SQL failed: {e}; sql: {sql}");
-            }
+            bg_client.simple_query(&sql).await.unwrap();
```
Comment (dentiny) [inline src/moonlink_connectors/src/pg_replicate/tests/cdc_event_test.rs]: I'm not sure unwrap would print out the sql statement, would appreciate if we use expect here

**Comment by peterxcli** [comment]
Thanks for the review!

File: src/moonlink_connectors/src/pg_replicate/tests/cdc_event_test.rs (hunk 29 above)
Comment (dentiny) [reply src/moonlink_connectors/src/pg_replicate/tests/cdc_event_test.rs]: FYI: a small followup https://github.com/Mooncake-Labs/moonlink/pull/1626

File: .github/workflows/ci.yml [hunk 30]
```diff
@@ -456,6 +457,18 @@ jobs:
           PGPASSWORD=postgres psql -h localhost -p 5432 -U postgres -d postgres -c "ALTER SYSTEM SET wal_level = 'logical';"
           PGPASSWORD=postgres psql -h localhost -p 5432 -U postgres -d postgres -c "SELECT pg_reload_conf();"
 
+      - name: Restart PostgreSQL to apply wal_level
+        run: |
```
Comment (dentiny) [inline .github/workflows/ci.yml:461]: oops, we need to set timeout to all operations; I met situations where minio / fake GCS setup gets stuck.
